bigquery/v2/query/query.go

```diff
@@ -0,0 +1,191 @@
+// Copyright 2025 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package query
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"cloud.google.com/go/bigquery/v2/apiv2/bigquerypb"
+	"github.com/googleapis/gax-go/v2"
+	"google.golang.org/protobuf/types/known/wrapperspb"
+)
+
+// Query represents a query job.
+type Query struct {
+	c         *Client
+	projectID string
+	jobID     string
+	location  string
+	queryID   string
+
+	// context for background pooling
+	ctx      context.Context
+	mu       sync.RWMutex
+	complete bool
+	ready    chan struct{}
+	err      error
+
+	cachedTotalRows uint64
+}
+
+// Create Query handler from jobs.query response and start background pooling job
+func newQueryJobFromQueryResponse(ctx context.Context, c *Client, res *bigquerypb.QueryResponse, opts ...gax.CallOption) (*Query, error) {
+	q := &Query{
+		c:       c,
+		ctx:     ctx,
+		queryID: res.QueryId,
+		ready:   make(chan struct{}),
+	}
+	err := q.consumeQueryResponse(&bigquerypb.GetQueryResultsResponse{
+		Schema:              res.Schema,
+		PageToken:           res.PageToken,
+		TotalRows:           res.TotalRows,
+		JobReference:        res.JobReference,
+		Rows:                res.Rows,
+		JobComplete:         res.JobComplete,
+		Errors:              res.Errors,
+		CacheHit:            res.CacheHit,
+		TotalBytesProcessed: res.TotalBytesProcessed,
+		NumDmlAffectedRows:  res.NumDmlAffectedRows,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	go q.waitForQueryBackground(opts)
+
+	return q, nil
+}
+
+// Create Query handler from JobReference response and start background pooling job
+func newQueryJobFromJobReference(ctx context.Context, c *Client, jobRef *bigquerypb.JobReference, opts ...gax.CallOption) (*Query, error) {
+	res := &bigquerypb.QueryResponse{
+		JobReference: jobRef,
+	}
+	return newQueryJobFromQueryResponse(ctx, c, res, opts...)
+}
+
+// Read returns a RowIterator for the query results.
+func (q *Query) Read(ctx context.Context, opts ...ReadOption) (*RowIterator, error) {
+	// TODO: proper setup iterator
+	return &RowIterator{}, nil
+}
+
+// Wait waits for the query to complete.
```

Comment:
Per our earlier discussion, the question here is whether to simply surface a done channel idom?